home/tests/test_content_import_export.py

```diff
@@ -1,109 +1,487 @@
+class ImportExportFixture:
+    admin_client: Any
+    importer: str
+    format: str
+
+    @property
+    def _import_content(self) -> Callable[..., None]:
+        return {
+            "new": import_content,
+            "old": old_import_content,
+        }[self.importer]
+
+    @property
+    def _filter_export(self) -> Callable[..., bytes]:
+        return {
+            "csv": self._filter_export_CSV,
+            "xlsx": self._filter_export_XLSX,
+        }[self.format]
+
+    def _filter_export_row(self, row: ExpDict, locale: str | None) -> bool:
+        """
+        Determine whether to keep a given export row.
+        """
+        if locale:
+            if row["locale"] not in [None, "", locale]:
+                return False
+        return True
+
+    def _filter_export_CSV(self, content: bytes, locale: str | None) -> bytes:
+        reader = csv.DictReader(StringIO(content.decode()))
+        assert reader.fieldnames is not None
+        out_content = StringIO()
+        writer = csv.DictWriter(out_content, fieldnames=reader.fieldnames)
+        writer.writeheader()
+        for row in reader:
+            if self._filter_export_row(row, locale=locale):
+                writer.writerow(row)
+        return out_content.getvalue().encode()
+
+    def _filter_export_XLSX(self, content: bytes, locale: str | None) -> bytes:
+        workbook = load_workbook(BytesIO(content))
+        worksheet = workbook.worksheets[0]
+        header = next(worksheet.iter_rows(max_row=1, values_only=True))
+
+        rows_to_remove: list[int] = []
+        for i, row in enumerate(worksheet.iter_rows(min_row=2, values_only=True)):
+            r = {k: v for k, v in zip(header, row, strict=True) if isinstance(k, str)}
+            if not self._filter_export_row(r, locale=locale):
+                rows_to_remove.append(i + 2)
+        for row_num in reversed(rows_to_remove):
+            worksheet.delete_rows(row_num)
+
+        out_content = BytesIO()
+        workbook.save(out_content)
+        return out_content.getvalue()
+
+    def export_content(self, locale: str | None = None) -> bytes:
+        """
+        Export all (or filtered) content in the configured format.
+
+        FIXME:
+         * If we filter the export by locale, we only get ContentPage entries
+           for the given language, but we still get ContentPageIndex rows for
+           all languages.
+        """
+        url = f"/admin/home/contentpage/?export={self.format}"
+        if locale:
+            loc = Locale.objects.get(language_code=locale)
+            locale = str(loc)
+            url = f"{url}&locale__id__exact={loc.id}"
+        content = self.admin_client.get(url).content
+        # Hopefully we can get rid of this at some point.
+        if locale:
+            content = self._filter_export(content, locale=locale)
+        if self.format == "csv":
+            print("-v-CONTENT-v-")
+            print(content.decode())
+            print("-^-CONTENT-^-")
```

Comment:
Do we still need this output for the tests?

Reply:
I added and removed it twice, so the third time I just left it in. 🤷

It's useful for figuring out whether it's the import or the export that's causing a test to fail.